src/evaluate/references.ts

```diff
@@ -0,0 +1,108 @@
+import * as cdk from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+
+/**
+ * A referenceable record.
+ *
+ * If `attributes` is set, the references is a map of sort and can return an attribute value for a given key.
+ */
+export interface Reference {
+  logicalId: string;
+  ref: string;
+  instance?: any;
+  hasAtt(attribute: string): boolean;
+  hasProp(property: string): boolean;
+}
+
+export class References {
+  private _store: Map<string, Reference> = new Map();
+
+  public add(reference: Reference): void {
+    this._store.set(reference.logicalId, reference);
+  }
+
+  public get(logicalId: string): Reference | undefined {
+    return this._store.get(logicalId);
+  }
+}
+
+export class SimpleReference implements Reference {
+  public constructor(public readonly logicalId: string) {}
+
+  public get ref(): string {
+    return this.logicalId;
+  }
+
+  public hasAtt(_attribute: string) {
+    return false;
+  }
+
+  public hasProp(_property: string) {
+    return false;
+  }
+}
+
+export class InstanceReference extends SimpleReference {
+  public constructor(logicalId: string, public readonly instance: any) {
+    super(logicalId);
+  }
+}
+
+export class ValueOnlyReference extends InstanceReference {
+  public get ref(): string {
+    throw new Error(`Ref: ${this.logicalId} cannot be referenced`);
```

Comment:
Is this necessarily true? A lazy resource can return a construct (e.g., `aws_lambda.Function.addAlias`).

Reply:
Fixed.